bot/cogs/free.py

```diff
@@ -0,0 +1,127 @@
+                msg = messages[seek-1]
+            # Otherwise get last message
+            else:
+                msg = await channel.history(limit=1).next()   # noqa (False positive)
+
+            inactive = (datetime.utcnow() - msg.created_at).seconds
+            if inactive > TIMEOUT:
+                free_channels.append((inactive, channel))
+
+        embed = Embed()
+        embed.colour = Colour.blurple()
+        embed.title = "**Looking for a free help channel?**"
+
+        if user is not None:
+            embed.description = f"**Hey {user.mention}!**\n\n"
+        else:
+            embed.description = ""
+
+        # Display all potentially inactive channels
+        # in descending order of inactivity
+        if free_channels:
+            embed.description += "**The following channel{0} look{1} free:**\n\n**".format(
+                's' if len(free_channels) > 1 else '',
+                '' if len(free_channels) > 1 else 's'
+            )
+
+            # Sort channels in descending order by seconds
+            # Get position in list, inactivity, and channel object
+            # For each channel, add to embed.description
+            for i, (inactive, channel) in enumerate(sorted(free_channels, reverse=True), 1):
+                minutes, seconds = divmod(inactive, 60)
+                if minutes > 59:
+                    hours, minutes = divmod(minutes, 60)
+                    embed.description += f"{i}. {channel.mention} inactive for {hours}h{minutes}m{seconds}s\n\n"
+                else:
+                    embed.description += f"{i}. {channel.mention} inactive for {minutes}m{seconds}s\n\n"
+
+            embed.description += ("**\nThese channels aren't guaranteed to be free, "
+                                  "so use your best judgement and check for yourself.")
+        else:
+            embed.description = ("**Doesn't look like any channels are available right now. "
+                                 "You're welcome to check for yourself to be sure. "
+                                 "If all channels are truly busy, please be patient "
+                                 "as one will likely be available soon.**")
+
+        await ctx.send(embed=embed)
+
+    @free.error
+    async def free_error(self, ctx: Context, error):
+        """
+        If error raised is CommandOnCooldown, and the
+        user who invoked has the helper role, reset
+        the cooldown and reinvoke the command.
+
+        Otherwise log the error.
+        """
+        helpers = ctx.guild.get_role(Roles.helpers)
+
+        if isinstance(error, commands.CommandOnCooldown):
+            if helpers in ctx.author.roles:
+                # reset cooldown so second invocation
+                # doesn't bring us back here.
+                ctx.command.reset_cooldown(ctx)
+                # return to avoid needlessly logging the error
+                return await ctx.invoke(ctx.command)
```

Comment:
haha dammit, I was debating with myself over which one was correct to use. I'm wondering if this sentence in the docs means I should not reset the cooldown first?

>This is similar to invoke() except that it bypasses checks, cooldowns, and error handlers.

My thinking is that it if a helper invokes this command with the cooldown active, they would reset the cooldown for everyone else since when the command is reinvoked, presumably the invocation from the helper wouldn't count towards the cooldown, allowing normal users to then use the command right after a helper since it would have been reset to zero just beforehand. Thus a user could use the command, then a helper directly after, and then a user could use the command right after that, having the command invoked 3 times within a short period of time.

Obviously I don't see this issue cropping up, but if I'm right and the cooldown doesn't increment during a `reinvoke`, then I might as well remove it.

Reply:
That's an interesting point. I did not consider that.

However, I did some testing and found that a regular user (non-helper) can call `!free` after a helper has bypassed the cooldown with the error function, regardless of whether `invoke` or `reinvoke` was used in the error function. It seem like these two functions may differ in their ability to bypass checks, but that they both don't start a new cooldown period.

So, the cooldown doesn't work after a bypass, as a new period doesn't get triggered by the `invoke` and `reinvoke` commands. (See below, this is with your `ctx.invoke` version) That means we either need to code a custom cooldown system that doesn't suffer from that or accept that it may happen.

I lean to the latter at the moment (accept it), as I think a custom cooldown system is an entire project in itself and should be reusable throughout the code. It does mean we can just use `reinvoke` to propagate the arguments.

![no_cooldown_for_nonhelper](https://user-images.githubusercontent.com/33516116/51078555-0332d500-16b7-11e9-88de-2f3c7034311e.png)

Reply:
When we get around to starting our `bot-core` project properly, a better (universal) cooldown system might be a good fit for that. For now, I think I'll resolve this.

Reply:
If you're going to resolve, I'd recommend us using `ctx.reinvoke` to pass the parameters instead of `ctx.invoke` since it doesn't matter for the cooldown, but does for the parameters getting passed.

Reply:
yeah you're right. let's do that @fiskenslakt

Reply:
They should be able to commit from the suggest code above, I think. That's what the suggestion feature is supposed to do. Saves another push and may keep your review from going stale.